File: frontend/src/api/reports.ts
import api from './axios'
import type {
  Activity,
  AuditLogListResponse,
  ImportPreviewResponse,
  NotificationPrefs,
  Opportunity,
} from '../types'

export async function downloadExcel(path: string, filename: string) {
  const res = await api.get(path, { responseType: 'blob' })
  const url = window.URL.createObjectURL(res.data)
  const link = document.createElement('a')
  link.href = url
  link.download = filename
  link.click()
  window.URL.revokeObjectURL(url)
}

export async function downloadPdf(path: string, filename: string) {
  const res = await api.get(path, { responseType: 'blob' })
  const url = window.URL.createObjectURL(res.data)
  const link = document.createElement('a')
  link.href = url
  link.download = filename
  link.click()
  window.URL.revokeObjectURL(url)
}

function queryString(params?: Record<string, string | undefined>) {
  const qs = new URLSearchParams()
  if (params) {
    for (const [key, value] of Object.entries(params)) {
      if (value) qs.set(key, value)
    }
  }
  const s = qs.toString()
  return s ? `?${s}` : ''
}

export const reportsApi = {
  opportunities: (params?: Record<string, string | undefined>) =>
    api.get<{ items: Opportunity[]; total: number }>('/reports/opportunities', { params }),

  activities: (params?: Record<string, string | undefined>) =>
    api.get<{ items: Activity[]; total: number }>('/reports/activities', { params }),

  exportAccounts: (template = false) =>
    downloadExcel(
      `/reports/export/accounts${template ? '?template=true' : ''}`,
      template ? 'accounts_template.xlsx' : 'accounts.xlsx'
    ),

  exportContacts: (template = false) =>
    downloadExcel(
      `/reports/export/contacts${template ? '?template=true' : ''}`,
      template ? 'contacts_template.xlsx' : 'contacts.xlsx'
    ),

  exportOpportunities: () => downloadExcel('/reports/export/opportunities', 'opportunities.xlsx'),

  exportActivities: () => downloadExcel('/reports/export/activities', 'activities.xlsx'),

  exportOpportunitiesPdf: (params?: Record<string, string | undefined>) =>
    downloadPdf(`/reports/export/pdf/opportunities${queryString(params)}`, 'opportunities_report.pdf'),

  exportActivitiesPdf: (params?: Record<string, string | undefined>) =>
    downloadPdf(`/reports/export/pdf/activities${queryString(params)}`, 'activities_report.pdf'),

  previewImportAccounts: (file: File) => {
    const form = new FormData()
    form.append('file', file)
    return api.post<ImportPreviewResponse>('/reports/import/accounts/preview', form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
  },

  confirmImportAccounts: (records: Record<string, unknown>[]) =>
    api.post<{ created_count: number; message: string }>('/reports/import/accounts/confirm', {
      records,
    }),

  previewImportContacts: (file: File) => {
    const form = new FormData()
    form.append('file', file)
    return api.post<ImportPreviewResponse>('/reports/import/contacts/preview', form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
  },

  confirmImportContacts: (records: Record<string, unknown>[]) =>
    api.post<{ created_count: number; message: string }>('/reports/import/contacts/confirm', {
      records,
    }),

  auditLog: (params?: Record<string, string | number | undefined>) =>
    api.get<AuditLogListResponse>('/reports/audit-log', { params }),
}

export type { NotificationPrefs }

File: backend/app/routers/reports.py
from datetime import datetime
from urllib.parse import quote

from fastapi import APIRouter, Depends, File, Query, UploadFile
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import get_current_user, require_manager
from app.models.activity import ActivityType
from app.models.audit_log import AuditAction
from app.models.opportunity import SalesStage
from app.models.user import User
from app.schemas.reports import (
    AuditLogListResponse,
    ImportConfirmRequest,
    ImportConfirmResponse,
    ImportPreviewResponse,
)
from app.services import reports_service
from app.services.exceptions import ServiceError
from app.services.http import raise_http

router = APIRouter()


def _excel_response(buffer, filename: str) -> StreamingResponse:
    encoded = quote(filename)
    return StreamingResponse(
        buffer,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f"attachment; filename*=UTF-8''{encoded}"},
    )


def _pdf_response(buffer, filename: str) -> StreamingResponse:
    encoded = quote(filename)
    return StreamingResponse(
        buffer,
        media_type="application/pdf",
        headers={"Content-Disposition": f"attachment; filename*=UTF-8''{encoded}"},
    )


@router.get("/opportunities")
async def report_opportunities(
    stage: SalesStage | None = Query(None),
    account_id: str | None = Query(None),
    assigned_to_id: str | None = Query(None),
    search: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    return await reports_service.report_opportunities(
        db,
        current_user,
        stage=stage,
        account_id=account_id,
        assigned_to_id=assigned_to_id,
        search=search,
    )


@router.get("/activities")
async def report_activities(
    account_id: str | None = Query(None),
    opportunity_id: str | None = Query(None),
    activity_type: ActivityType | None = Query(None),
    assigned_to: str | None = Query(None),
    from_date: datetime | None = Query(None),
    to_date: datetime | None = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    return await reports_service.report_activities(
        db,
        current_user,
        account_id=account_id,
        opportunity_id=opportunity_id,
        activity_type=activity_type,
        assigned_to=assigned_to,
        from_date=from_date,
        to_date=to_date,
    )


@router.get("/export/accounts")
async def export_accounts(
    template: bool = Query(False),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    buffer = await reports_service.export_accounts(db, template=template)
    name = "accounts_template.xlsx" if template else "accounts.xlsx"
    return _excel_response(buffer, name)


@router.get("/export/contacts")
async def export_contacts(
    template: bool = Query(False),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    buffer = await reports_service.export_contacts(db, template=template)
    name = "contacts_template.xlsx" if template else "contacts.xlsx"
    return _excel_response(buffer, name)


@router.get("/export/opportunities")
async def export_opportunities(
    template: bool = Query(False),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    buffer = await reports_service.export_opportunities(db, template=template)
    name = "opportunities_template.xlsx" if template else "opportunities.xlsx"
    return _excel_response(buffer, name)


@router.get("/export/activities")
async def export_activities(
    template: bool = Query(False),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    buffer = await reports_service.export_activities(db, template=template)
    name = "activities_template.xlsx" if template else "activities.xlsx"
    return _excel_response(buffer, name)


@router.post("/import/accounts/preview", response_model=ImportPreviewResponse)
async def preview_import_accounts(
    file: UploadFile = File(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    try:
        content = await file.read()
        return await reports_service.preview_import_accounts(db, content)
    except ServiceError as e:
        raise_http(e)


@router.post("/import/accounts/confirm", response_model=ImportConfirmResponse)
async def confirm_import_accounts(
    body: ImportConfirmRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    try:
        return await reports_service.confirm_import_accounts(db, current_user, body.records)
    except ServiceError as e:
        raise_http(e)


@router.post("/import/contacts/preview", response_model=ImportPreviewResponse)
async def preview_import_contacts(
    file: UploadFile = File(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    try:
        content = await file.read()
        return await reports_service.preview_import_contacts(db, content)
    except ServiceError as e:
        raise_http(e)


@router.post("/import/contacts/confirm", response_model=ImportConfirmResponse)
async def confirm_import_contacts(
    body: ImportConfirmRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    try:
        return await reports_service.confirm_import_contacts(db, current_user, body.records)
    except ServiceError as e:
        raise_http(e)


@router.get("/export/pdf/opportunities")
async def export_opportunities_pdf(
    stage: SalesStage | None = Query(None),
    account_id: str | None = Query(None),
    assigned_to_id: str | None = Query(None),
    search: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    buffer = await reports_service.export_opportunities_pdf(
        db,
        current_user,
        stage=stage,
        account_id=account_id,
        assigned_to_id=assigned_to_id,
        search=search,
    )
    return _pdf_response(buffer, "opportunities_report.pdf")


@router.get("/export/pdf/activities")
async def export_activities_pdf(
    account_id: str | None = Query(None),
    opportunity_id: str | None = Query(None),
    activity_type: ActivityType | None = Query(None),
    assigned_to: str | None = Query(None),
    from_date: datetime | None = Query(None),
    to_date: datetime | None = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    buffer = await reports_service.export_activities_pdf(
        db,
        current_user,
        account_id=account_id,
        opportunity_id=opportunity_id,
        activity_type=activity_type,
        assigned_to=assigned_to,
        from_date=from_date,
        to_date=to_date,
    )
    return _pdf_response(buffer, "activities_report.pdf")


@router.get("/audit-log", response_model=AuditLogListResponse)
async def get_audit_log(
    entity_type: str | None = Query(None),
    user_id: str | None = Query(None),
    action: AuditAction | None = Query(None),
    from_date: datetime | None = Query(None),
    to_date: datetime | None = Query(None),
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_manager),
):
    return await reports_service.list_audit_logs(
        db,
        entity_type=entity_type,
        user_id=user_id,
        action=action,
        from_date=from_date,
        to_date=to_date,
        page=page,
        per_page=per_page,
    )
